smoke-simplest e2e test fails intermittently

I have seen this on CI and can reproduce it on my laptop by running the test repeatedly.  The last time I did so it failed on the 11th iteration with the output below.  The key part is:  ".spec.ports.name: value mismatch, expected: jaeger-grpc != actual: otlp-grpc" as if it's confusing the jaeger-grpc and otlp-grpc ports.

```
=== RUN   kuttl/harness/smoke-simplest
=== PAUSE kuttl/harness/smoke-simplest
=== CONT  kuttl/harness/smoke-simplest
    logger.go:42: 10:29:06 | smoke-simplest | Creating namespace: kuttl-test-kind-kangaroo
    logger.go:42: 10:29:06 | smoke-simplest/0-install | starting test step 0-install
I0712 10:29:08.218699   29266 request.go:655] Throttling request took 1.045197889s, request: GET:https://192.168.64.2:8443/apis/discovery.k8s.io/v1beta1?timeout=32s
    logger.go:42: 10:29:08 | smoke-simplest/0-install | OpenTelemetryCollector:kuttl-test-kind-kangaroo/simplest created
    logger.go:42: 10:31:39 | smoke-simplest/0-install | test step failed 0-install
    case.go:361: failed in step 0-install
    case.go:363: --- Service:kuttl-test-kind-kangaroo/simplest-collector
        +++ Service:kuttl-test-kind-kangaroo/simplest-collector
        @@ -1,15 +1,85 @@
         apiVersion: v1
         kind: Service
         metadata:
        +  labels:
        +    app.kubernetes.io/component: opentelemetry-collector
        +    app.kubernetes.io/instance: kuttl-test-kind-kangaroo.simplest
        +    app.kubernetes.io/managed-by: opentelemetry-operator
        +    app.kubernetes.io/name: simplest-collector
        +    app.kubernetes.io/part-of: opentelemetry
        +    app.kubernetes.io/version: latest
        +  managedFields:
        +  - apiVersion: v1
        +    fieldsType: FieldsV1
        +    fieldsV1:
        +      f:metadata:
        +        f:labels:
        +          .: {}
        +          f:app.kubernetes.io/component: {}
        +          f:app.kubernetes.io/instance: {}
        +          f:app.kubernetes.io/managed-by: {}
        +          f:app.kubernetes.io/name: {}
        +          f:app.kubernetes.io/part-of: {}
        +          f:app.kubernetes.io/version: {}
        +        f:ownerReferences:
        +          .: {}
        +          k:{"uid":"dfc52c93-ed85-4cab-8953-0a02fe04c545"}: {}
        +      f:spec:
        +        f:internalTrafficPolicy: {}
        +        f:ports:
        +          .: {}
        +          k:{"port":4317,"protocol":"TCP"}:
        +            .: {}
        +            f:appProtocol: {}
        +            f:name: {}
        +            f:port: {}
        +            f:protocol: {}
        +            f:targetPort: {}
        +          k:{"port":4318,"protocol":"TCP"}:
        +            .: {}
        +            f:appProtocol: {}
        +            f:name: {}
        +            f:port: {}
        +            f:protocol: {}
        +            f:targetPort: {}
        +          k:{"port":14250,"protocol":"TCP"}:
        +            .: {}
        +            f:appProtocol: {}
        +            f:name: {}
        +            f:port: {}
        +            f:protocol: {}
        +            f:targetPort: {}
        +          k:{"port":55681,"protocol":"TCP"}:
        +            .: {}
        +            f:appProtocol: {}
        +            f:name: {}
        +            f:port: {}
        +            f:protocol: {}
        +            f:targetPort: {}
        +        f:selector: {}
        +        f:sessionAffinity: {}
        +        f:type: {}
        +    manager: manager
        +    operation: Update
        +    time: "2022-07-12T08:29:08Z"
           name: simplest-collector
           namespace: kuttl-test-kind-kangaroo
        +  ownerReferences:
        +  - apiVersion: opentelemetry.io/v1alpha1
        +    blockOwnerDeletion: true
        +    controller: true
        +    kind: OpenTelemetryCollector
        +    name: simplest
        +    uid: dfc52c93-ed85-4cab-8953-0a02fe04c545
         spec:
        +  clusterIP: 10.111.179.189
        +  clusterIPs:
        +  - 10.111.179.189
        +  internalTrafficPolicy: Cluster
        +  ipFamilies:
        +  - IPv4
        +  ipFamilyPolicy: SingleStack
           ports:
        -  - appProtocol: grpc
        -    name: jaeger-grpc
        -    port: 14250
        -    protocol: TCP
        -    targetPort: 14250
           - appProtocol: grpc
             name: otlp-grpc
             port: 4317
        @@ -25,4 +95,20 @@
             port: 55681
             protocol: TCP
             targetPort: 4318
        +  - appProtocol: grpc
        +    name: jaeger-grpc
        +    port: 14250
        +    protocol: TCP
        +    targetPort: 14250
        +  selector:
        +    app.kubernetes.io/component: opentelemetry-collector
        +    app.kubernetes.io/instance: kuttl-test-kind-kangaroo.simplest
        +    app.kubernetes.io/managed-by: opentelemetry-operator
        +    app.kubernetes.io/name: simplest-collector
        +    app.kubernetes.io/part-of: opentelemetry
        +    app.kubernetes.io/version: latest
        +  sessionAffinity: None
        +  type: ClusterIP
        +status:
        +  loadBalancer: {}


    case.go:363: resource Service:kuttl-test-kind-kangaroo/simplest-collector: .spec.ports.name: value mismatch, expected: jaeger-grpc != actual: otlp-grpc
Warning: events.k8s.io/v1beta1 Event is deprecated in v1.22+, unavailable in v1.25+
    logger.go:42: 10:31:39 | smoke-simplest | smoke-simplest events from ns kuttl-test-kind-kangaroo:
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:09 +0200 CEST	Normal	Pod simplest-collector-85b4bc9b74-g6b8f		Scheduled	Successfully assigned kuttl-test-kind-kangaroo/simplest-collector-85b4bc9b74-g6b8f to minikube
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:09 +0200 CEST	Normal	Pod simplest-collector-85b4bc9b74-g6b8f.spec.containers{otc-container}		Pulled	Container image "ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector:0.54.0" already present on machine
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:09 +0200 CEST	Normal	Pod simplest-collector-85b4bc9b74-g6b8f.spec.containers{otc-container}		Created	Created container otc-container
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:09 +0200 CEST	Normal	ReplicaSet.apps simplest-collector-85b4bc9b74		SuccessfulCreate	Created pod: simplest-collector-85b4bc9b74-g6b8f
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:09 +0200 CEST	Normal	Deployment.apps simplest-collector		ScalingReplicaSet	Scaled up replica set simplest-collector-85b4bc9b74 to 1
    logger.go:42: 10:31:39 | smoke-simplest | 2022-07-12 10:29:10 +0200 CEST	Normal	Pod simplest-collector-85b4bc9b74-g6b8f.spec.containers{otc-container}		Started	Started container otc-container
=== CONT  kuttl
    harness.go:402: run tests finished
    harness.go:511: cleaning up
    harness.go:568: removing temp folder: ""
--- FAIL: kuttl (165.93s)
    --- FAIL: kuttl/harness (0.00s)
        --- FAIL: kuttl/harness/smoke-simplest (152.86s)
FAIL
```

--- a/pkg/collector/adapters/config_to_ports.go
+++ b/pkg/collector/adapters/config_to_ports.go
@@ -62,8 +62,13 @@ func ConfigToReceiverPorts(logger logr.Logger, config map[interface{}]interface{
 	}
 
 	sortedNames := make([]string, 0, len(receivers))
-	for name := range receivers {
-		sortedNames = append(sortedNames, name.(string))
+	for receiverName := range receivers {
+		name, ok := receiverName.(string)
+		if !ok {
+			logger.V(2).Info("receiver does not match type, want: %s, got: %T", "string", receiverName)
+			continue
+		}
+		sortedNames = append(sortedNames, name)
 	}
 	sort.Strings(sortedNames)
 

--- a/pkg/collector/adapters/config_to_ports.go
+++ b/pkg/collector/adapters/config_to_ports.go
@@ -16,6 +16,7 @@ package adapters
 
 import (
 	"errors"
+	"sort"
 
 	"github.com/go-logr/logr"
 	corev1 "k8s.io/api/core/v1"
@@ -60,28 +61,34 @@ func ConfigToReceiverPorts(logger logr.Logger, config map[interface{}]interface{
 		return nil, ErrReceiversNotAMap
 	}
 
+	sortedNames := make([]string, 0, len(receivers))
+	for name := range receivers {
+		sortedNames = append(sortedNames, name.(string))
+	}
+	sort.Strings(sortedNames)
+
 	ports := []corev1.ServicePort{}
-	for key, val := range receivers {
+	for _, name := range sortedNames {
 		// This check will pass only the enabled receivers,
 		// then only the related ports will be opened.
-		if !recEnabled[key] {
+		if !recEnabled[name] {
 			continue
 		}
-		receiver, ok := val.(map[interface{}]interface{})
+
+		receiver, ok := receivers[name].(map[interface{}]interface{})
 		if !ok {
-			logger.Info("receiver doesn't seem to be a map of properties", "receiver", key)
+			logger.Info("receiver doesn't seem to be a map of properties", "receiver", name)
 			receiver = map[interface{}]interface{}{}
 		}
 
-		rcvrName := key.(string)
-		rcvrParser := parser.For(logger, rcvrName, receiver)
+		rcvrParser := parser.For(logger, name, receiver)
 
 		rcvrPorts, err := rcvrParser.Ports()
 		if err != nil {
 			// should we break the process and return an error, or just ignore this faulty parser
 			// and let the other parsers add their ports to the service? right now, the best
 			// option seems to be to log the failures and move on, instead of failing them all
-			logger.Error(err, "parser for '%s' has returned an error: %w", rcvrName, err)
+			logger.Error(err, "parser for '%s' has returned an error: %w", name, err)
 			continue
 		}
 

--- a/pkg/collector/adapters/config_to_ports.go
+++ b/pkg/collector/adapters/config_to_ports.go
@@ -61,48 +61,41 @@ func ConfigToReceiverPorts(logger logr.Logger, config map[interface{}]interface{
 		return nil, ErrReceiversNotAMap
 	}
 
-	sortedNames := make([]string, 0, len(receivers))
-	for receiverName := range receivers {
-		name, ok := receiverName.(string)
-		if !ok {
-			logger.V(2).Info("receiver does not match type, want: %s, got: %T", "string", receiverName)
-			continue
-		}
-		sortedNames = append(sortedNames, name)
-	}
-	sort.Strings(sortedNames)
-
 	ports := []corev1.ServicePort{}
-	for _, name := range sortedNames {
+	for key, val := range receivers {
 		// This check will pass only the enabled receivers,
 		// then only the related ports will be opened.
-		if !recEnabled[name] {
+		if !recEnabled[key] {
 			continue
 		}
-
-		receiver, ok := receivers[name].(map[interface{}]interface{})
+		receiver, ok := val.(map[interface{}]interface{})
 		if !ok {
-			logger.Info("receiver doesn't seem to be a map of properties", "receiver", name)
+			logger.Info("receiver doesn't seem to be a map of properties", "receiver", key)
 			receiver = map[interface{}]interface{}{}
 		}
 
-		rcvrParser := parser.For(logger, name, receiver)
+		rcvrName := key.(string)
+		rcvrParser := parser.For(logger, rcvrName, receiver)
 
 		rcvrPorts, err := rcvrParser.Ports()
 		if err != nil {
 			// should we break the process and return an error, or just ignore this faulty parser
 			// and let the other parsers add their ports to the service? right now, the best
 			// option seems to be to log the failures and move on, instead of failing them all
-			logger.Error(err, "parser for '%s' has returned an error: %w", name, err)
+			logger.Error(err, "parser for '%s' has returned an error: %w", rcvrName, err)
 			continue
 		}
 
 		if len(rcvrPorts) > 0 {
-			sort.Slice(ports, func(i, j int) bool {
-				return ports[i].Name < ports[j].Name
-			})
 			ports = append(ports, rcvrPorts...)
 		}
 	}
+
+	if len(ports) > 0 {
+		sort.Slice(ports, func(i, j int) bool {
+			return ports[i].Name < ports[j].Name
+		})
+	}
+
 	return ports, nil
 }

--- a/pkg/collector/adapters/config_to_ports.go
+++ b/pkg/collector/adapters/config_to_ports.go
@@ -98,6 +98,9 @@ func ConfigToReceiverPorts(logger logr.Logger, config map[interface{}]interface{
 		}
 
 		if len(rcvrPorts) > 0 {
+			sort.Slice(ports, func(i, j int) bool {
+				return ports[i].Name < ports[j].Name
+			})
 			ports = append(ports, rcvrPorts...)
 		}
 	}

--- a/pkg/collector/adapters/config_to_ports.go
+++ b/pkg/collector/adapters/config_to_ports.go
@@ -91,11 +91,9 @@ func ConfigToReceiverPorts(logger logr.Logger, config map[interface{}]interface{
 		}
 	}
 
-	if len(ports) > 0 {
-		sort.Slice(ports, func(i, j int) bool {
-			return ports[i].Name < ports[j].Name
-		})
-	}
+	sort.Slice(ports, func(i, j int) bool {
+		return ports[i].Name < ports[j].Name
+	})
 
 	return ports, nil
 }

--- a/pkg/collector/adapters/config_to_ports_test.go
+++ b/pkg/collector/adapters/config_to_ports_test.go
@@ -144,6 +144,39 @@ service:
 	}
 }
 
+func TestReceiverPortsSortedCorrectly(t *testing.T) {
+	configurationString := `receivers:
+  otlp:
+    protocols:
+      grpc:
+      http:  
+  jaeger:
+    protocols:
+      grpc:
+service:
+  pipelines:
+    traces:
+      receivers: [otlp, jaeger]
+      exporters: [logging]
+`
+	configuration, err := adapters.ConfigFromString(configurationString)
+	require.NoError(t, err)
+	require.NotEmpty(t, configuration)
+
+	ports, err := adapters.ConfigToReceiverPorts(logger, configuration)
+	assert.NoError(t, err)
+	assert.Len(t, ports, 4)
+
+	// Ports should be ordered by receiver (alphabetically) and by port number within receiver
+	assert.Equal(t, "jaeger-grpc", ports[0].Name)
+	assert.Equal(t, int32(14250), ports[0].Port)
+
+	assert.Equal(t, "otlp-grpc", ports[1].Name)
+	assert.Equal(t, int32(4317), ports[1].Port)
+	assert.Equal(t, int32(4318), ports[2].Port)
+	assert.Equal(t, int32(55681), ports[3].Port)
+}
+
 func TestNoPortsParsed(t *testing.T) {
 	for _, tt := range []struct {
 		desc      string

--- a/pkg/collector/adapters/config_to_ports_test.go
+++ b/pkg/collector/adapters/config_to_ports_test.go
@@ -22,6 +22,7 @@ import (
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	logf "sigs.k8s.io/controller-runtime/pkg/log"
 
 	"github.com/open-telemetry/opentelemetry-operator/pkg/collector/adapters"
@@ -87,94 +88,24 @@ service:
 	assert.Len(t, ports, 11)
 
 	// verify
-	expectedPorts := map[int32]bool{}
-	expectedPorts[int32(12345)] = false
-	expectedPorts[int32(12346)] = false
-	expectedPorts[int32(14250)] = false
-	expectedPorts[int32(6831)] = false
-	expectedPorts[int32(6833)] = false
-	expectedPorts[int32(15268)] = false
-	expectedPorts[int32(4318)] = false
-	expectedPorts[int32(55681)] = false
-	expectedPorts[int32(55555)] = false
-	expectedPorts[int32(9411)] = false
+	httpAppProtocol := "http"
+	grpcAppProtocol := "grpc"
+	targetPortZero := intstr.IntOrString{Type: 0, IntVal: 0, StrVal: ""}
+	targetPort4317 := intstr.IntOrString{Type: 0, IntVal: 4317, StrVal: ""}
+	targetPort4318 := intstr.IntOrString{Type: 0, IntVal: 4318, StrVal: ""}
 
-	expectedNames := map[string]bool{}
-	expectedNames["examplereceiver"] = false
-	expectedNames["examplereceiver-settings"] = false
-	expectedNames["jaeger-grpc"] = false
-	expectedNames["jaeger-thrift-compact"] = false
-	expectedNames["jaeger-thrift-binary"] = false
-	expectedNames["jaeger-custom-thrift-http"] = false
-	expectedNames["otlp-grpc"] = false
-	expectedNames["otlp-http"] = false
-	expectedNames["otlp-http-legacy"] = false
-	expectedNames["otlp-2-grpc"] = false
-	expectedNames["zipkin"] = false
-
-	expectedAppProtocols := map[string]string{}
-	expectedAppProtocols["otlp-grpc"] = "grpc"
-	expectedAppProtocols["otlp-http"] = "http"
-	expectedAppProtocols["otlp-http-legacy"] = "http"
-	expectedAppProtocols["jaeger-custom-thrift-http"] = "http"
-	expectedAppProtocols["jaeger-grpc"] = "grpc"
-	expectedAppProtocols["otlp-2-grpc"] = "grpc"
-	expectedAppProtocols["zipkin"] = "http"
-
-	// make sure we only have the ports in the set
-	for _, port := range ports {
-		assert.NotNil(t, expectedPorts[port.Port])
-		assert.NotNil(t, expectedNames[port.Name])
-		expectedPorts[port.Port] = true
-		expectedNames[port.Name] = true
-
-		if appProtocol, ok := expectedAppProtocols[port.Name]; ok {
-			assert.Equal(t, appProtocol, *port.AppProtocol)
-		}
-	}
-
-	// and make sure all the ports from the set are there
-	for _, val := range expectedPorts {
-		assert.True(t, val)
-	}
-
-	// make sure we only have the ports names in the set
-	for _, val := range expectedNames {
-		assert.True(t, val)
-	}
-}
-
-func TestReceiverPortsSortedCorrectly(t *testing.T) {
-	configurationString := `receivers:
-  otlp:
-    protocols:
-      grpc:
-      http:  
-  jaeger:
-    protocols:
-      grpc:
-service:
-  pipelines:
-    traces:
-      receivers: [otlp, jaeger]
-      exporters: [logging]
-`
-	configuration, err := adapters.ConfigFromString(configurationString)
-	require.NoError(t, err)
-	require.NotEmpty(t, configuration)
-
-	ports, err := adapters.ConfigToReceiverPorts(logger, configuration)
-	assert.NoError(t, err)
-	assert.Len(t, ports, 4)
-
-	// Ports should be ordered by receiver (alphabetically) and by port number within receiver
-	assert.Equal(t, "jaeger-grpc", ports[0].Name)
-	assert.Equal(t, int32(14250), ports[0].Port)
-
-	assert.Equal(t, "otlp-grpc", ports[1].Name)
-	assert.Equal(t, int32(4317), ports[1].Port)
-	assert.Equal(t, int32(4318), ports[2].Port)
-	assert.Equal(t, int32(55681), ports[3].Port)
+	assert.Len(t, ports, 11)
+	assert.Equal(t, corev1.ServicePort{Name: "examplereceiver", Port: int32(12345)}, ports[0])
+	assert.Equal(t, corev1.ServicePort{Name: "examplereceiver-settings", Port: int32(12346)}, ports[1])
+	assert.Equal(t, corev1.ServicePort{Name: "jaeger-custom-thrift-http", AppProtocol: &httpAppProtocol, Protocol: "TCP", Port: int32(15268), TargetPort: targetPortZero}, ports[2])
+	assert.Equal(t, corev1.ServicePort{Name: "jaeger-grpc", AppProtocol: &grpcAppProtocol, Protocol: "TCP", Port: int32(14250)}, ports[3])
+	assert.Equal(t, corev1.ServicePort{Name: "jaeger-thrift-binary", Protocol: "UDP", Port: int32(6833)}, ports[4])
+	assert.Equal(t, corev1.ServicePort{Name: "jaeger-thrift-compact", Protocol: "UDP", Port: int32(6831)}, ports[5])
+	assert.Equal(t, corev1.ServicePort{Name: "otlp-2-grpc", AppProtocol: &grpcAppProtocol, Protocol: "TCP", Port: int32(55555)}, ports[6])
+	assert.Equal(t, corev1.ServicePort{Name: "otlp-grpc", AppProtocol: &grpcAppProtocol, Port: int32(4317), TargetPort: targetPort4317}, ports[7])
+	assert.Equal(t, corev1.ServicePort{Name: "otlp-http", AppProtocol: &httpAppProtocol, Port: int32(4318), TargetPort: targetPort4318}, ports[8])
+	assert.Equal(t, corev1.ServicePort{Name: "otlp-http-legacy", AppProtocol: &httpAppProtocol, Port: int32(55681), TargetPort: targetPort4318}, ports[9])
+	assert.Equal(t, corev1.ServicePort{Name: "zipkin", AppProtocol: &httpAppProtocol, Protocol: "TCP", Port: int32(9411)}, ports[10])
 }
 
 func TestNoPortsParsed(t *testing.T) {

